internal/search/search.go

```diff
@@ -0,0 +1,220 @@
+		}
+	}
+
+	for i, line := range context {
+		context[i] = truncateLine(line)
+	}
+
+	ret := ld.HunkRep{
+		ProjKey:            projKey,
+		FlagKey:            flagKey,
+		StartingLineNumber: startingLineNum + 1,
+		Lines:              strings.Join(context, "\n"),
+		Aliases:            []string{},
+	}
+	ret.Aliases = helpers.Dedupe(append(ret.Aliases, aliasMatches...))
+	return &ret
+}
+
+// aggregateHunksForFlag finds all references in a file, and combines matches if their context lines overlap
+func (f file) aggregateHunksForFlag(projKey, flagKey string, flagAliases []string, ctxLines int, delimiters string) []ld.HunkRep {
+	hunksForFlag := []ld.HunkRep{}
+	for i := range f.lines {
+		match := f.hunkForLine(projKey, flagKey, flagAliases, i, ctxLines, delimiters)
+		if match != nil {
+			lastHunkIdx := len(hunksForFlag) - 1
+			// If the previous hunk overlaps or is adjacent to the current hunk, merge them together
+			if lastHunkIdx >= 0 && hunksForFlag[lastHunkIdx].Overlap(*match) >= 0 {
+				hunksForFlag = append(hunksForFlag[:lastHunkIdx], mergeHunks(hunksForFlag[lastHunkIdx], *match)...)
+			} else {
+				hunksForFlag = append(hunksForFlag, *match)
+			}
+		}
+	}
+	return hunksForFlag
+}
+
+func (f file) toHunks(projKey string, aliases map[string][]string, ctxLines int, delimiters string) *ld.ReferenceHunksRep {
+	hunks := []ld.HunkRep{}
+	for flagKey, flagAliases := range aliases {
+		hunks = append(hunks, f.aggregateHunksForFlag(projKey, flagKey, flagAliases, ctxLines, delimiters)...)
+	}
+	if len(hunks) == 0 {
+		return nil
+	}
+	return &ld.ReferenceHunksRep{Path: f.path, Hunks: hunks}
+}
+
+// mergeHunks combines the lines and aliases of two hunks together for a given file
+// if the hunks do not overlap, returns each hunk separately
+// assumes the startingLineNumber of a is less than b and there is some overlap between the two
+func mergeHunks(a, b ld.HunkRep) []ld.HunkRep {
+	if a.StartingLineNumber > b.StartingLineNumber {
+		a, b = b, a
+	}
+
+	aLines := strings.Split(a.Lines, "\n")
+	bLines := strings.Split(b.Lines, "\n")
+	overlap := a.Overlap(b)
+	// no overlap
+	if overlap < 0 || len(a.Lines) == 0 && len(b.Lines) == 0 {
+		return []ld.HunkRep{a, b}
+	} else if overlap >= len(bLines) {
+		// subset hunk
+		return []ld.HunkRep{a}
+	}
+
+	combinedLines := append(aLines, bLines[overlap:]...)
+	return []ld.HunkRep{
+		{
+			StartingLineNumber: a.StartingLineNumber,
+			Lines:              strings.Join(combinedLines, "\n"),
+			ProjKey:            a.ProjKey,
+			FlagKey:            a.FlagKey,
+			Aliases:            helpers.Dedupe(append(a.Aliases, b.Aliases...)),
+		},
+	}
+}
+
+// processFiles starts goroutines to process files individually. When all files have completed processing, the references channel is closed to signal completion.
+func processFiles(files chan file, references chan ld.ReferenceHunksRep, projKey string, aliases map[string][]string, ctxLines int, delimiters string) {
```

Comment:
for safety you can specify the channel direction:
https://gobyexample.com/channel-directions